xds/internal/xdsclient/transport/ads/ads_stream.go

```diff
@@ -0,0 +1,780 @@
+// is sufficient flow control quota.
+func (s *Stream) Subscribe(typ xdsresource.Type, name string) {
+	if s.logger.V(2) {
+		s.logger.Infof("Subscribing to resource %q of type %q", name, typ.TypeName())
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	state, ok := s.resourceTypeState[typ]
+	if !ok {
+		// An entry in the type state map is created as part of the first
+		// subscription request for this type.
+		state = &resourceTypeState{
+			subscribedResources: make(map[string]*ResourceWatchState),
+			bufferedRequests:    make(chan struct{}, 1),
+		}
+		s.resourceTypeState[typ] = state
+	}
+
+	// Create state for the newly subscribed resource. The watch timer will
+	// be started when a request for this resource is actually sent out.
+	state.subscribedResources[name] = &ResourceWatchState{State: ResourceWatchStateStarted}
+
+	// Send a request for the resource type with updated subscriptions.
+	s.requestCh.Put(typ)
+}
+
+// Unsubscribe cancels the subscription to the given resource. It is a no-op if
+// the given resource does not exist. The watch expiry timer associated with the
+// resource is stopped if one is active. A discovery request is sent out on the
+// stream for the resource type when there is sufficient flow control quota.
+func (s *Stream) Unsubscribe(typ xdsresource.Type, name string) {
+	if s.logger.V(2) {
+		s.logger.Infof("Unsubscribing to resource %q of type %q", name, typ.TypeName())
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	state, ok := s.resourceTypeState[typ]
+	if !ok {
+		s.logger.Infof("easwars: returning early while Unsubscribing to resource %q of type %q", name, typ.TypeName())
+		return
+	}
+
+	rs, ok := state.subscribedResources[name]
+	if !ok {
+		s.logger.Infof("easwars: returning early while Unsubscribing to resource %q of type %q", name, typ.TypeName())
+		return
+	}
+	if rs.ExpiryTimer != nil {
+		rs.ExpiryTimer.Stop()
+	}
+	delete(state.subscribedResources, name)
+
+	// Send a request for the resource type with updated subscriptions.
+	s.requestCh.Put(typ)
+}
+
+// runner is a long-running goroutine that handles the lifecycle of the ADS
+// stream. It creates a new stream when the previous one fails, and sends
+// discovery requests on the stream. It backs off before creating new streams
+// following stream failures.
+func (s *Stream) runner(ctx context.Context) {
+	defer close(s.runnerDoneCh)
+
+	go s.send(ctx)
+
+	runStreamWithBackoff := func() error {
+		stream, err := s.transport.CreateStreamingCall(ctx, "/envoy.service.discovery.v3.AggregatedDiscoveryService/StreamAggregatedResources")
+		if err != nil {
+			s.logger.Warningf("Creating new ADS stream failed: %v", err)
+			s.onError(err, false)
+			return nil
+		}
+		if s.logger.V(2) {
+			s.logger.Infof("ADS stream created")
+		}
+
+		s.mu.Lock()
+		// Flow control is a property of the underlying stream and needs to be
+		// initialized everytime a new stream is created.
+		s.fc = newADSFlowControl()
+		s.firstRequest = true
+		s.mu.Unlock()
```

Comment:
I don't get this protection of this critical section in this goroutine. If the previous stream is failed and the backoff goes off, how can these fields be accessed? You can call operations on a failed stream (the new steram hasn't been sent on the channel yet, so it'll never be accessed)? What are the races here?

Reply:
Oh I see, these fields are to the whole ads stream type, whereas the stream local variable is this one iteration. So you already spawned the send goroutine, it can access these fields, on send path while also concurrent Subscribe/Unsubscribe. I thought this was on the local stream which won't get accessed on send until you send it later.